src/rgw/rgw_rest_s3.cc

```diff
@@ -4589,13 +4589,14 @@ int RGWHandler_REST_S3::init_from_header(rgw::sal::RGWRadosStore *store,
   if (s->init_state.url_bucket.empty()) {
     // Save bucket to tide us over until token is parsed.
     s->init_state.url_bucket = first;
+    string encoded_obj_str;
     if (pos >= 0) {
-      string encoded_obj_str = req.substr(pos+1);
-      if (s->bucket) {
-	s->object = s->bucket->get_object(rgw_obj_key(encoded_obj_str, s->info.args.get("versionId")));
-      } else {
-	s->object = store->get_object(rgw_obj_key(encoded_obj_str, s->info.args.get("versionId")));
-      }
+      encoded_obj_str = req.substr(pos+1);
+    }
+    if (s->bucket) {
+      s->object = s->bucket->get_object(rgw_obj_key(encoded_obj_str, s->info.args.get("versionId")));
```

Comment:
This doesn't work, since encoded_obj_str is only set when pos>=0.  It needs to stay inside the above if().

Reply:
In the case without the last slash, segfault occurs because s->object is accessed without being set. If the path ends with the last slash, the empty object is set to s->object. I thought the case without the last slash need to be the same.

Reply:
If it is fixed as you commented, it will look like this:
```
diff --git a/src/rgw/rgw_rest_s3.cc b/src/rgw/rgw_rest_s3.cc
index bb856aa..99071cd 100644
--- a/src/rgw/rgw_rest_s3.cc
+++ b/src/rgw/rgw_rest_s3.cc
@@ -4857,7 +4857,9 @@ RGWHandler_REST* RGWRESTMgr_S3::get_handler(rgw::sal::RGWRadosStore *store,
 }

 bool RGWHandler_REST_S3Website::web_dir() const {
-  std::string subdir_name = url_decode(s->object->get_name());
+  std::string subdir_name;
+  if (s->object)
+    subdir_name = url_decode(s->object->get_name());

   if (subdir_name.empty()) {
     return false;
@@ -4917,7 +4919,8 @@ int RGWHandler_REST_S3Website::retarget(RGWOp* op, RGWOp** new_op) {
   }

   rgw_obj_key new_obj;
-  bool get_res = s->bucket->get_info().website_conf.get_effective_key(s->object->get_name(), &new_obj.name, web_dir());
+  bool get_res = s->bucket->get_info().website_conf.get_effective_key(
+      s->object ? s->object->get_name() : "", &new_obj.name, web_dir());
   if (!get_res) {
     s->err.message = "The IndexDocument Suffix is not configurated or not well formed!";
     ldpp_dout(s, 5) << s->err.message << dendl;
```
But I'm still wondering why the empty name cannot be used in the case without the last slash, even though the empty name is already used in the case with the last slash.

Reply:
Hmm... I missed that if there is a / but no object name, then it will create an empty object.  I'm not sure what to make of this.  @cbodley any thoughts?

Reply:
So, we talked this over during the upstream meeting today, and decided that zipper objects shouldn't be empty, but should be null instead.  So something like this:
https://github.com/dang/ceph/commit/83770df68db52e2e632196b47d0cba4d2a38d7a3

Reply:
If you'd prefer, I can submit that, or you can fix this one up.

Reply:
No problem. It's okay to submit your PR.

Reply:
Okay, please go review https://github.com/ceph/ceph/pull/37975